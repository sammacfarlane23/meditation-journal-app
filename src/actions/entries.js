import database from '../firebase/firebase';

export const createEntry = (entry) => ({
  type: 'CREATE_ENTRY',
  entry,
});

export const startCreateEntry = (entryData = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    const { entryText = '', createdAt = 0, title = '' } = entryData;
    const entry = {
      entryText,
      createdAt,
      title,
    };

    return database
      .ref(`users/${uid}/entries`)
      .push(entry)
      .then((ref) => {
        dispatch(
          createEntry({
            id: ref.key,
            ...entry,
          })
        );
      });
  };
};

export const removeEntry = ({ id } = {}) => ({
  type: 'REMOVE_ENTRY',
  id,
});

export const startRemoveEntry = ({ id } = {}) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .remove()
      .then(() => {
        dispatch(removeEntry({ id }));
      });
  };
};

export const editEntry = (id, updates) => ({
  type: 'EDIT_ENTRY',
  id,
  updates,
});

export const startEditEntry = (id, updates) => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editEntry(id, updates));
      });
  };
};

export const setEntries = (entries) => {
  return {
    type: 'SET_ENTRIES',
    entries,
  };
};

export const startSetEntries = () => {
  return (dispatch, getState) => {
    const uid = getState().auth.uid;
    return database
      .ref(`users/${uid}/entries`)
      .once('value')
      .then((snapshot) => {
        const entries = [];

        snapshot.forEach((childSnapshot) => {
          entries.push({
            id: childSnapshot.key,
            ...childSnapshot.val(),
          });
        });
        dispatch(setEntries(entries));
      });
  };
};

export const selectEntryForEdit = (entry) => {
  return {
    type: 'SELECT_ENTRY_FOR_EDIT',
    entry,
  };
};
