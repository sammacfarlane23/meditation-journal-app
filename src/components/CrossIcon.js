import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';

export default (props) => (
  <div>
    <button className='footer-button' onClick={props.showSearch}>
      <FontAwesomeIcon icon={faTimes} className='white-button' size='2x' />
    </button>
  </div>
);
