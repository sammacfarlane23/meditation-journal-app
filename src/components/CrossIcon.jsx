import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faTimes } from '@fortawesome/free-solid-svg-icons';

export default (props) => (
  <div>
    <button className='button--big' onClick={props.showSearch}>
      <FontAwesomeIcon icon={faTimes} size='2x' />
    </button>
  </div>
);
