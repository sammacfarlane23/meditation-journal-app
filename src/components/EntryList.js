import React from 'react';
import { connect } from 'react-redux';
import EntryItem from './EntryItem';
import selectEntries from '../selectors/entries';

export const EntryList = (props) => (
  <div className='entry-list'>
    {props.entries.length === 0 ? (
      <div className='entry-item entry-item--message'>
        <span>No entries</span>
      </div>
    ) : (
      props.entries.map((entry) => {
        return <EntryItem key={entry.id} {...entry} />;
      })
    )}
  </div>
);

const mapStateToProps = (state) => {
  return {
    entries: selectEntries(state.entries, state.filters),
  };
};

export default connect(mapStateToProps)(EntryList);
