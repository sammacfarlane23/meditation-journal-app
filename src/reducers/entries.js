const entriesReducerDefaultState = [];

export default (state = entriesReducerDefaultState, action) => {
  switch (action.type) {
    case 'CREATE_ENTRY':
      return [...state, action.entry];
    case 'REMOVE_ENTRY':
      return state.filter(({ id }) => action.id !== id);
    case 'EDIT_ENTRY':
      return state.map((entry) => {
        if (entry.id === action.id) {
          return {
            ...entry,
            ...action.updates,
          };
        } else {
          return entry;
        }
      });
    case 'SET_ENTRIES':
      return action.entries;
    case 'SELECT_ENTRY_FOR_EDIT':
      return action.entry;
    default:
      return state;
  }
};
