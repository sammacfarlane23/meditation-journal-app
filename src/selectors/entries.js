export default (entries, { text }) => {
  return entries
    .filter((entry) => {
      const titleMatch = entry.title.toLowerCase().includes(text.toLowerCase());

      const textMatch = entry.entryText
        .toLowerCase()
        .includes(text.toLowerCase());

      return textMatch || titleMatch;
    })
    .sort((a, b) => {
      return a.createdAt < b.createdAt ? 1 : -1;
    });
};
