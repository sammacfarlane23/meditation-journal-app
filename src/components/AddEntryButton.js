import React, { useState } from 'react';
import { useDispatch } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus } from '@fortawesome/free-solid-svg-icons';
import EntryModal from './EntryModal';
import { startCreateEntry } from '../actions/entries';

export default () => {
  const [isModalOpen, setIsModalOpen] = useState(false);
  const dispatch = useDispatch();

  const showModal = () => {
    setIsModalOpen(true);
  };

  const onSubmit = (entry) => {
    dispatch(startCreateEntry(entry));
    setIsModalOpen(false);
  };

  const closeModal = () => {
    setIsModalOpen(false);
  };

  return (
    <div className='button-position'>
      {!isModalOpen && (
        <button className='button--big circular-container' onClick={showModal}>
          <FontAwesomeIcon icon={faPlus} size='2x' />
        </button>
      )}
      <EntryModal
        showModal={isModalOpen}
        onSubmit={onSubmit}
        closeModal={closeModal}
      />
    </div>
  );
};
