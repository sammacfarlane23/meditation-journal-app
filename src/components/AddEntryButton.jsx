import React from 'react';
import { connect } from 'react-redux';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faPlus } from '@fortawesome/free-solid-svg-icons';
import EntryModal from './EntryModal';
import { startCreateEntry } from '../actions/entries';

export class AddEntryButton extends React.Component {
  state = { show: false };

  showModal = () => {
    this.setState(() => ({ show: true }));
  };

  onSubmit = (entry) => {
    this.props.startCreateEntry(entry);
    this.setState({ show: false });
  };

  closeModal = () => {
    this.setState({ show: false });
  };
  render() {
    return (
      <div className='button-position'>
        {!this.state.show && (
          <button
            className='button--big circular-container'
            onClick={this.showModal}
          >
            <FontAwesomeIcon icon={faPlus} size='2x' />
          </button>
        )}
        <EntryModal
          showModal={this.state.show}
          onSubmit={this.onSubmit}
          closeModal={this.closeModal}
        />
      </div>
    );
  }
}

const mapDispatchToProps = (dispatch) => ({
  startCreateEntry: (entry) => dispatch(startCreateEntry(entry)),
});

export default connect(undefined, mapDispatchToProps)(AddEntryButton);
