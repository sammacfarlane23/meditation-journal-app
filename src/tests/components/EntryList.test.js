import React from 'react';
import { shallow } from 'enzyme';
import { EntryList } from '../../components/EntryList';
import entries from '../fixtures/entries';

test('should render EntryList with entries', () => {
  const wrapper = shallow(<EntryList hasEntries={true} entries={entries} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render welcome message if no entries', () => {
  const wrapper = shallow(<EntryList hasEntries={false} entries={[]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render no entries message if there are entries but none selected', () => {
  const wrapper = shallow(<EntryList hasEntries={true} entries={[]} />);
  expect(wrapper).toMatchSnapshot();
});
