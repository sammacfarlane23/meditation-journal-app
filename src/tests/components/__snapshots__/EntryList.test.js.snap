// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EntryList with entries 1`] = `
<div
  className="entry-list"
>
  <[object Object]
    createdAt={0}
    entryText="This is an entry"
    id="1"
    key="1"
    title=""
  />
  <[object Object]
    createdAt={-1000}
    entryText="This is another entry"
    id="2"
    key="2"
    title="title"
  />
  <[object Object]
    createdAt={1600000000000}
    entryText="This is third entry"
    id="3"
    key="3"
    title=""
  />
</div>
`;

exports[`should render no entries message if there are entries but none selected 1`] = `
<div
  className="entry-list"
>
  <div
    className="entry-item entry-item--message"
  >
    <span>
      No matching entries :(
    </span>
  </div>
</div>
`;

exports[`should render welcome message if no entries 1`] = `
<div
  className="entry-list"
>
  <div
    className="entry-item entry-item--message"
  >
    Welcome to Meditation Journal App. Click the plus button below to add an entry.
  </div>
</div>
`;
